api/client/client_test.go

```diff
@@ -0,0 +1,487 @@
+	}
+	if person.ID == "" {
+		t.Fatalf("expected an entity id to be set but it wasn't")
+	}
+
+	has, err := txn.Has(existingPerson.ID)
+	if err != nil {
+		t.Fatalf("failed to write txn has: %v", err)
+	}
+	if !has {
+		t.Fatalf("expected has to be true but it wasn't")
+	}
+
+	foundExistingPerson := &Person{}
+	err = txn.FindByID(existingPerson.ID, foundExistingPerson)
+	if err != nil {
+		t.Fatalf("failed to txn find by id: %v", err)
+	}
+	if !reflect.DeepEqual(foundExistingPerson, existingPerson) {
+		t.Fatalf("txn model found by id does't equal the original")
+	}
+
+	existingPerson.Age = 99
+	err = txn.Save(existingPerson)
+	if err != nil {
+		t.Fatalf("failed to save in write txn: %v", err)
+	}
+
+	err = txn.Delete(existingPerson.ID)
+	if err != nil {
+		t.Fatalf("failed to delete in write txn: %v", err)
+	}
+}
+
+func TestListen(t *testing.T) {
+	_, clean := server(t)
+	defer clean()
+	client := client(t)
+
+	storeID, err := client.NewStore()
+	checkErr(t, err)
+	err = client.RegisterSchema(storeID, modelName, schema)
+	checkErr(t, err)
+	err = client.Start(storeID)
+	checkErr(t, err)
+
+	person := createPerson()
+
+	err = client.ModelCreate(storeID, modelName, person)
+	checkErr(t, err)
+
+	channel, discard, err := client.Listen(storeID, modelName, person.ID)
+	if err != nil {
+		t.Fatalf("failed to listen: %v", err)
+	}
+
+	go func() {
+		time.Sleep(1 * time.Second)
+		person.Age = 30
+		_ = client.ModelSave(storeID, modelName, person)
+		person.Age = 40
+		_ = client.ModelSave(storeID, modelName, person)
+	}()
+
+	val, ok := <-channel
+	if !ok {
+		t.Fatal("channel no longer active at first event")
+	} else {
+		p := &Person{}
+		if err := json.Unmarshal(val, p); err != nil {
+			t.Fatalf("failed to unmarshal listen result: %v", err)
+		}
+		if p.Age != 30 {
+			t.Fatalf("expected listen result age = 30 but got: %v", p.Age)
+		}
+	}
+
+	val, ok = <-channel
+	if !ok {
+		t.Fatal("channel no longer active at second event")
+	} else {
+		p := &Person{}
+		if err := json.Unmarshal(val, p); err != nil {
+			t.Fatalf("failed to unmarshal listen result: %v", err)
+		}
+		if p.Age != 40 {
+			t.Fatalf("expected listen result age = 40 but got: %v", p.Age)
+		}
+	}
+	discard()
+}
+
+func server(t *testing.T) (*api.Server, func()) {
+	dir, err := ioutil.TempDir("", "")
+	checkErr(t, err)
+	ts, err := es.DefaultThreadservice(
+		dir,
+		es.ListenPort(4006),
+		es.ProxyPort(5050),
+		es.Debug(true))
+	checkErr(t, err)
+	ts.Bootstrap(util.DefaultBoostrapPeers())
+	server, err := api.NewServer(context.Background(), ts, api.Config{
+		RepoPath: dir,
+		Debug:    true,
+	})
+	checkErr(t, err)
+	return server, func() {
+		server.Close()
+		if err := ts.Close(); err != nil {
+			panic(err)
+		}
+		os.RemoveAll(dir)
+	}
+}
+
+func client(t *testing.T) *Client {
+	client, err := NewClient("localhost", 9090)
+	checkErr(t, err)
+	return client
+}
+
+func checkErr(t *testing.T, err error) {
+	t.Helper()
+	if err != nil {
+		t.Fatal(err)
+	}
+}
```

Comment:
Strange thing... I can run each of these tests individually and the pass (other than `TestListen`), but if I comment out `TestListen` and try to run the package tests, it hangs and times out. No idea what could be causing that.

Reply:
Tracking this here https://github.com/textileio/go-textile-threads/issues/146